### PR TITLE
fix a race in purger managing in process delete requests

### PR DESCRIPTION
**Pull request by sandeepsukhani**
**What this PR does**:
Fixed a race in purger while managing in process delete requests.
Refactored some code to manage them easily.

**Checklist**
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/chunk/purger/purger.go [hunk 1]
```diff
@@ -454,10 +444,7 @@ func (p *Purger) pullDeleteRequestsToPlanDeletes() error {
 			return err
 		}
 
-		p.inProcessRequestIDsMtx.Lock()
-		p.inProcessRequests[deleteRequest.UserID] = deleteRequest
-		p.inProcessRequestIDsMtx.Unlock()
-
+		p.inProcessRequests.add(deleteRequest.UserID, &deleteRequest)
```
Comment (pstibrany) [inline pkg/chunk/purger/purger.go:447]: This takes address of variable declared in `for` loop. This address will be the same for all iterations.

For this to work correctly, copy of deleteRequest must be made:

```suggestion
		deleteRequestCopy := deleteRequest
		p.inProcessRequests.add(deleteRequest.UserID, &deleteRequestCopy)
```

File: pkg/chunk/purger/purger.go [hunk 2]
```diff
@@ -369,10 +366,10 @@ func (p *Purger) loadInprocessDeleteRequests() error {
 
 	for _, deleteRequest := range requestsWithBuildingPlanStatus {
 		req := makeDeleteRequestWithLogger(deleteRequest, util.Logger)
+		p.inProcessRequests.add(deleteRequest.UserID, &deleteRequest)
```
Comment (pstibrany) [inline pkg/chunk/purger/purger.go:369]: Same problem with using address of for-loop variable.

File: pkg/chunk/purger/purger.go [hunk 3]
```diff
@@ -393,7 +390,7 @@ func (p *Purger) loadInprocessDeleteRequests() error {
 		req := makeDeleteRequestWithLogger(deleteRequest, util.Logger)
 		level.Info(req.logger).Log("msg", "loaded in process delete requests with status deleting")
 
-		p.inProcessRequests[deleteRequest.UserID] = deleteRequest
+		p.inProcessRequests.add(deleteRequest.UserID, &deleteRequest)
```
Comment (pstibrany) [inline pkg/chunk/purger/purger.go:393]: Once again.

File: pkg/chunk/purger/purger.go (hunk 1 above)
Comment (sandeepsukhani) [reply pkg/chunk/purger/purger.go:447]: Thanks for the catch! I totally missed it.

File: pkg/chunk/purger/purger.go [hunk 4]
```diff
@@ -714,3 +691,56 @@ func makeDeleteRequestWithLogger(deleteRequest DeleteRequest, l log.Logger) dele
...
+type inProcessRequestsCollection struct {
+	requests map[string]*DeleteRequest
+	mtx      sync.RWMutex
+}
+
+func newInProcessRequestsCollection() *inProcessRequestsCollection {
+	return &inProcessRequestsCollection{requests: map[string]*DeleteRequest{}}
+}
+
+func (i *inProcessRequestsCollection) add(userID string, request *DeleteRequest) {
```
Comment (pracucci) [inline pkg/chunk/purger/purger.go:706]: I would call it `set()` because overrides any previous request (if any) and doesn't "add" the request to a list.

**Pull request merged**